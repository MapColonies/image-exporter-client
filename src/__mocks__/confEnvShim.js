if (!window._env_) {
  window._env_ = (function (undefined) {
    return {
      LANGUAGE: 'en',
      MAP_SERVER: 'MAP_SERVER',
      PUBLISH_POINT: 'PUBLISH_POINT',
      CHANNEL: 1002,
      VERSION: 1,
      REQUEST: 'REQUEST',
      SERVICE_PROTOCOL: 'SERVICE_PROTOCOL',
      SERVICE_NAME: 'SERVICE_NAME',
      ACTIVE_LAYER: 'ACTIVE_LAYER',
      BBOX: {
        limit: 20000,
      },
      EXPORT: {
        defaultZoomLevel: 18,
        maxZoomLevel: 22
      },
      MAP: {
        center: [35.14, 31.39],
        zoom: 8,
      },
      ACTIVE_LAYER_PROPERTIES: {
        urlPattern : 'MAP_SERVER/service',
        exportUrlPattern: 'MAP_SERVER/service?REQUEST=REQUEST&SERVICE=SERVICE&LAYERS=LAYER',
        urlPatternParams: {
          service: 'SERVICE',
          layers: 'LAYER',
          tiled: 'TILED',
          matrixSet: 'MATRIXSET',
          style: 'STYLE',
          projection: 'PROJECTION'
        }
      },
    };
  })(void 0);
}
