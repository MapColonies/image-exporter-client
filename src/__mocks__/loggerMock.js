export default {
  log: () => {},
  info: () => {},
  error: () => {},
  debug: () => {},
};
