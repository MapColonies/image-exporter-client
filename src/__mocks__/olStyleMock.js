export class Fill {
  constructor() { }
}

export class Stroke {
  constructor() { }
}

export class Circle {
  constructor() { }
}

export class Style {
  constructor() { }
}
