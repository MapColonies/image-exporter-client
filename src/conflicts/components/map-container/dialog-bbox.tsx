import React, { useState } from 'react';
import * as  turf from '@turf/helpers';
import bbox from '@turf/bbox';
import bboxPolygon from '@turf/bbox-polygon';
import distance  from '@turf/distance/dist/js'; //TODO: make a consumption "REGULAR"
import { Polygon } from 'geojson';
import { useFormik } from 'formik';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import { 
  Dialog, 
  DialogTitle, 
  DialogContent, 
  TextField, 
  Button} from '@map-colonies/react-core';
import { Box } from '@map-colonies/react-components';
import EXPORTER_CONFIG from '../../../common/config';

  const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    bbox:{
      overflow: 'hidden',
      position: 'relative',
      border: 'solid 1px', 
      width: '60px', 
      height: '56px',
    },

    bboxLeftBottomCorner: {
      "&::before": {
        margin: '-1em',
        borderRadius: '50%',
        position: 'absolute',
        padding: '1em',
        boxShadow: '0 0 7px #b53',
        background: '#95a',
        content: "''",
        bottom: 0   
      }
    },
    bboxRightTopCorner: {
      "&::before": {
        margin: '-1em',
        borderRadius: '50%',
        position: 'absolute',
        padding: '1em',
        boxShadow: '0 0 7px #b53',
        background: '#95a',
        content: "''",
        top: 0,
        right: 0   
      }
    },
    spacer: {
      marginRight: '16px'
    },
    errorContainer: {
      display: 'flex',
      alignItems: 'center',
      marginRight: 'auto',
      color: theme.palette.error.main,
    }

  })
);

interface BBoxCorners {
  bottomLeftLat: number;
  bottomLeftLon: number;
  topRightLat: number;
  topRightLon: number;
}

interface BBoxCornersError {
  latDistance: string;
  lonDistance: string;
}

const validate = (values: BBoxCorners): BBoxCornersError => {
  const errors: BBoxCornersError = {latDistance: '', lonDistance: ''};

  try{
    turf.lineString([
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.topRightLat], 
    ]);

    const yDistance = distance(
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.bottomLeftLon, values.topRightLat]);
    
    const xDistance = distance(
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.bottomLeftLat]);
  
    if (xDistance > EXPORTER_CONFIG.BOUNDARIES.MAX_X_KM){
      errors.latDistance = 'X distance is exceeded the limit';
    }
    if (yDistance > EXPORTER_CONFIG.BOUNDARIES.MAX_Y_KM){
      errors.lonDistance = 'Y distance is exceeded the limit';
    }
  }
  catch(err){
    errors.latDistance = 'Not valid coordinates';
  }

  return errors;
};


interface DialogBBoxProps {
  isOpen: boolean;
  onSetOpen: (open: boolean) => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const DialogBBox: React.FC<DialogBBoxProps> = (
  props
) => {
  const { isOpen, onSetOpen, onPolygonUpdate } = props;
  const classes = useStyle();
  const formik = useFormik({
    initialValues: {
      bottomLeftLat: 0,
      bottomLeftLon: 0,
      topRightLat: 0,
      topRightLon: 0,
    },
    onSubmit: values => {
      const err = validate(values);
      if(!err.latDistance && !err.lonDistance){
        const line = turf.lineString([
          [values.bottomLeftLon, values.bottomLeftLat],
          [values.topRightLon, values.topRightLat], 
        ]);
        const polygon = bboxPolygon(bbox(line));
        console.log('polygon', polygon.geometry);
  
        onPolygonUpdate(polygon.geometry);
        handleClose(false);
      }
      else{
        setFormErrors(err);
      }

    },
  });

  const [formErrors, setFormErrors] = useState({latDistance: '', lonDistance: ''});
  
  const handleClose = (isOpened:boolean):void => {
    onSetOpen(isOpened);
  }
  return (
      <Dialog open={isOpen}>
        <DialogTitle>Define bbox coordinates</DialogTitle>
        <DialogContent>
        <form onSubmit={formik.handleSubmit}>
          <Box style={{display: 'flex', marginBottom: '16px'}}>
            <TextField 
              label="Top Right Lat" 
              id="topRightLat"
              name="topRightLat"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.topRightLat}
              className={classes.spacer}
            />
            <TextField 
              label="Top Right Lon" 
              id="topRightLon"
              name="topRightLon"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.topRightLon}
              className={classes.spacer}
            />
            <div className={`${classes.bbox} ${classes.bboxRightTopCorner}`}></div>
          </Box>
          <Box style={{display: 'flex'}}>
          <TextField 
              label="Bottom Left Lat" 
              id="bottomLeftLat"
              name="bottomLeftLat"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.bottomLeftLat}
              className={classes.spacer}
            />
            <TextField 
              label="Bottom Left Lon" 
              id="bottomLeftLon"
              name="bottomLeftLon"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.bottomLeftLon}
              className={classes.spacer}
            />
            <div className={`${classes.bbox} ${classes.bboxLeftBottomCorner}`}></div>
          </Box>
          <Box style={{ display: 'flex', justifyContent: 'flex-end', marginTop: '16px', gap: '16px' }}>
            {
              (!!formErrors.latDistance || !!formErrors.lonDistance) ? 
              <div className={classes.errorContainer}>
                {`Error: ${formErrors.latDistance} ${formErrors.lonDistance}`}
              </div> : 
              null
            }
            <Button type="button" onClick={()=>{handleClose(false);}}>Cancel</Button>
            <Button raised type="submit">Ok</Button>
          </Box>
        </form>
        </DialogContent>
      </Dialog>

  );
}
