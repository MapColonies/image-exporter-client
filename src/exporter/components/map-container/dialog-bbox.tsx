import React, { useState } from 'react';
import * as  turf from '@turf/helpers';
import bbox from '@turf/bbox';
import bboxPolygon from '@turf/bbox-polygon';
import { Polygon } from 'geojson';
import { useFormik } from 'formik';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  TextField,
  Button
} from '@map-colonies/react-core';
import { Box } from '@map-colonies/react-components';
import { FormattedMessage, useIntl, IntlShape } from 'react-intl';
import { BBoxCorner, Corner } from '../bbox/bbox-corner-indicator';
import { BBoxAreaLimit, isBBoxWithinLimit } from '../../../common/helpers/bbox-area';
import EXPORTER_CONFIG from '../../../common/config';

const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    spacer: {
      marginRight: '16px'
    },
    errorContainer: {
      display: 'flex',
      alignItems: 'center',
      marginRight: 'auto',
      color: theme.palette.error.main,
    }

  })
);

interface BBoxCorners {
  bottomLeftLat: number;
  bottomLeftLon: number;
  topRightLat: number;
  topRightLon: number;
}

interface BBoxError {
  bboxArea: string;
}

const validate = (values: BBoxCorners, intl: IntlShape): BBoxError => {
  const errors: BBoxError = { bboxArea: '' };

  try {
    const line = turf.lineString([
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.topRightLat],
    ]);
    const polygon = bboxPolygon(bbox(line));
    const isWithinLimit = isBBoxWithinLimit(polygon.geometry);
    if (isWithinLimit !== BBoxAreaLimit.OK) {
      const messageFormat = isWithinLimit === BBoxAreaLimit.TOO_BIG ? 'custom-bbox.form-error.area.large.text' : 'custom-bbox.form-error.area.small.text';
      errors.bboxArea = intl.formatMessage({ id: messageFormat });
    }
  } catch (err) {
    errors.bboxArea = intl.formatMessage({ id: 'custom-bbox.form-error.area.invalid.text' });
  }

  return errors;
};


interface DialogBBoxProps {
  isOpen: boolean;
  onSetOpen: (open: boolean) => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const DialogBBox: React.FC<DialogBBoxProps> = (
  props
) => {
  const { isOpen, onSetOpen, onPolygonUpdate } = props;
  const classes = useStyle();
  const intl = useIntl();
  const formik = useFormik({
    initialValues: {
      bottomLeftLat: 0,
      bottomLeftLon: 0,
      topRightLat: 0.00001,
      topRightLon: 0.00001,
    },
    onSubmit: values => {
      const err = validate(values, intl);

      if (!err.bboxArea) {
        const line = turf.lineString([
          [values.bottomLeftLon, values.bottomLeftLat],
          [values.topRightLon, values.topRightLat],
        ]);
        const polygon = bboxPolygon(bbox(line));

        onPolygonUpdate(polygon.geometry);
        handleClose(false);
        setFormErrors({
          bboxArea: ''
        });
      } else {
        setFormErrors(err);
      }

    },
  });

  const [formErrors, setFormErrors] = useState({ bboxArea: '' });

  const handleClose = (isOpened: boolean): void => {
    onSetOpen(isOpened);
  };

  // eslint-disable-next-line
  const checkValidLat = (e: React.ChangeEvent<any>) => {
    const maxMinLatValue = EXPORTER_CONFIG.MAP.MAX_MIN_LAT_VALUE;
    const isValid: boolean = isValidRegexPattern(e, maxMinLatValue);
    return isValid? formik.handleChange(e) : false;
  };

  // eslint-disable-next-line
  const checkValidLong = (e: React.ChangeEvent<any>) => {
    const maxMinLongValue = EXPORTER_CONFIG.MAP.MAX_MIN_LONG_VALUE;
    const isValid: boolean = isValidRegexPattern(e, maxMinLongValue);
    return isValid? formik.handleChange(e) : false;
  };

  const isValidRegexPattern = (e: React.ChangeEvent<any>, maxMinValue: number): boolean => {
    // eslint-disable-next-line
    const valueString: string = e.target.value as string;
    const splitChar = EXPORTER_CONFIG.EXPORT.SPLIT_CHARACTER;
    const maxFractionDigits = EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS;
    const regexStr = `^([0-9])+\\${splitChar}?(\\b[0-9]{0,${maxFractionDigits}})*$`;
    const regex = new RegExp(regexStr);
    const value = Number(valueString);
    if (!valueString) {
      return false;
    }
    if (value < -maxMinValue || value > maxMinValue) {
      return false;
    }
    return regex.test(valueString);
  };

  return (
    <Dialog open={ isOpen } preventOutsideDismiss={ true }>
      <DialogTitle>
        <FormattedMessage id="custom-bbox.dialog.title" />
      </DialogTitle>
      <DialogContent>
        <form onSubmit={ formik.handleSubmit }>
          <Box style={ { display: 'flex', marginBottom: '16px' } }>
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.top_right_lat.label' }) }
              id="topRightLat" 
              name="topRightLat"
              type="text"
              onChange={ checkValidLat }
              value={ formik.values.topRightLat }
              className={ classes.spacer }
            />
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.top_right_lon.label' }) }
              id="topRightLon"
              name="topRightLon"
              type="text"
              onChange={ checkValidLong }
              value={ formik.values.topRightLon }
              className={ classes.spacer }
            />
            <BBoxCorner corner={ Corner.TOP_RIGHT } />
          </Box>
          <Box style={ { display: 'flex' } }>
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.bottom_left_lat.label' }) }
              id="bottomLeftLat"
              name="bottomLeftLat"
              type="text"
              onChange={ checkValidLat }
              value={ formik.values.bottomLeftLat }
              className={ classes.spacer }
            />
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.bottom_left_lon.label' }) }
              id="bottomLeftLon"
              name="bottomLeftLon"
              type="text"
              onChange={ checkValidLong }
              value={ formik.values.bottomLeftLon }
              className={ classes.spacer }
            />
            <BBoxCorner corner={ Corner.BOTTOM_LEFT } />
          </Box>
          <Box style={ { display: 'flex', justifyContent: 'flex-end', marginTop: '16px', gap: '16px' } }>
            {
              formErrors.bboxArea ?
                <div id="errorContainer" className={ classes.errorContainer }>
                  { `${ intl.formatMessage({ id: 'general.error.label' }) }: ${ formErrors.bboxArea }` }
                </div> :
                null
            }
            <Button type="button" onClick={ (): void => {
              handleClose(false);
            } }>
              <FormattedMessage id="general.cancel-btn.text" />
            </Button>
            <Button raised type="submit">
              <FormattedMessage id="general.ok-btn.text" />
            </Button>
          </Box>
        </form>
      </DialogContent>
    </Dialog>

  );
};
