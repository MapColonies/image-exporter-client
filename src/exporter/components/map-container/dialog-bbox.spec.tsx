import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { act, waitFor } from '@testing-library/react';
import 'mutationobserver-shim';
import { TextField } from '@map-colonies/react-core';
import { getFieldValue } from '../../../common/test-helpers/text-field.helper.spec';
import { getButtonById } from '../../../common/test-helpers/button.helper.spec';
// eslint-disable-next-line
import '../../../__mocks__/confEnvShim';
import { DialogBBox } from './dialog-bbox';


global.MutationObserver = window.MutationObserver;

const setOpenFn = jest.fn();
const polygonUpdate = jest.fn();

// Enzyme doesnâ€™t work properly with hooks in general, especially for `shallow` so this is the way to mock `react-intl` module.
// Enspired by https://github.com/formatjs/formatjs/issues/1477
jest.mock('react-intl', () => {
  /* eslint-disable */
  const reactIntl = require.requireActual('react-intl');
  const MESSAGES = require.requireActual('../../../common/i18n');
  const intl = reactIntl.createIntl({
    locale: 'en',
    messages: MESSAGES.default['en'],
  });

  return {
    ...reactIntl,
    useIntl: () => intl,
  };
  /* eslint-enable */
});

const updateField = (wrapper: ShallowWrapper, fieldName: string, value: number) => {
  const fieldWrapper = wrapper.find(TextField).find({ name: fieldName });
  
  act(() => {
    fieldWrapper.at(0).simulate('change', {
      // simulate changing e.target.name and e.target.value
      target: {
        name: fieldName,
        value
      },
    });
  });

  act(() => {
    fieldWrapper.simulate('blur');
  });
};

describe('DialogBBox component', () => {

  afterEach(() => {
    setOpenFn.mockClear();
    polygonUpdate.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <DialogBBox
        isOpen={false}
        onSetOpen={setOpenFn}
        onPolygonUpdate={polygonUpdate}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('Bbox fields filled succesfully', async () => {
    const fields = {
      bottomLeftLat: '10.1',
      bottomLeftLon: '10.1',
      topRightLat: '10.2',
      topRightLon: '10.2',
    };

    const wrapper = shallow(
      <DialogBBox
        isOpen={false}
        onSetOpen={setOpenFn}
        onPolygonUpdate={polygonUpdate}
      />
    );

    for (const field in fields) {
      // eslint-disable-next-line
      updateField(wrapper, field, (fields as any)[field]);
    }

    await waitFor(() => {
      for (const field in fields) {
        // eslint-disable-next-line
        expect(getFieldValue(wrapper,field)).toEqual((fields as any)[field]);
      }
    });
  });

  it('Cancel button triggers dialog closing', () => {
    const wrapper = shallow(
      <DialogBBox
        isOpen={false}
        onSetOpen={setOpenFn}
        onPolygonUpdate={polygonUpdate}
      />
    );

    getButtonById(wrapper, 'general.cancel-btn.text').simulate('click');

    expect(setOpenFn).toHaveBeenCalledWith(false);
  });

  it('Submiting the form triggers onPolygonUpdate', async () => {
    const wrapper = shallow(
      <DialogBBox
        isOpen={false}
        onSetOpen={setOpenFn}
        onPolygonUpdate={polygonUpdate}
      />
    ); 

    act(()=>{
      wrapper
        .find('form')
        .simulate('submit');
    });

    wrapper.update();

    await waitFor(() => {
      expect(polygonUpdate).toHaveBeenCalled();
    });
  });

  //// TODO test to check error presentation logic.
  //// When component uses useFormik() hook internal formik state not updated when triggered onChange event.
  //// Probably formik should be used in component way rather than hook.
  //// Also preferable to test such functionality on the levevl of E2E tests.
  // it('Bbox stays opened and shows ERROR when Y(km) limit is exceeded', async () => {
  //   const fields = {
  //     bottomLeftLat: 31,
  //     bottomLeftLon: 34,
  //     topRightLat: 32,
  //     topRightLon: 35,
  //   };
  //   .
  //   .
  //   .  
  //   console.log('ERROR --->', wrapper.find('#errorContainer').length);
  //   expect(wrapper.find('#errorContainer')).toBeDefined();
  // });

});
