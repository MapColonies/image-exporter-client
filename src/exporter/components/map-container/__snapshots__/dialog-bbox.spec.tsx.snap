// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogBBox component renders correctly 1`] = `
<Function
  open={false}
  preventOutsideDismiss={true}
>
  <Function>
    <FormattedMessage
      id="custom-bbox.dialog.title"
    />
  </Function>
  <Function>
    <form
      onSubmit={[Function]}
    >
      <Component
        style={
          Object {
            "display": "flex",
            "marginBottom": "16px",
          }
        }
      >
        <Function
          className="makeStyles-spacer-1"
          id="topRightLat"
          label="Top Right Lat"
          name="topRightLat"
          onChange={[Function]}
          type="text"
          value={0.00001}
        />
        <Function
          className="makeStyles-spacer-1"
          id="topRightLon"
          label="Top Right Lon"
          name="topRightLon"
          onChange={[Function]}
          type="text"
          value={0.00001}
        />
        <BBoxCorner
          corner="TOP_RIGHT"
        />
      </Component>
      <Component
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <Function
          className="makeStyles-spacer-1"
          id="bottomLeftLat"
          label="Bottom Left Lat"
          name="bottomLeftLat"
          onChange={[Function]}
          type="text"
          value={0}
        />
        <Function
          className="makeStyles-spacer-1"
          id="bottomLeftLon"
          label="Bottom Left Lon"
          name="bottomLeftLon"
          onChange={[Function]}
          type="text"
          value={0}
        />
        <BBoxCorner
          corner="BOTTOM_LEFT"
        />
      </Component>
      <Component
        style={
          Object {
            "display": "flex",
            "gap": "16px",
            "justifyContent": "flex-end",
            "marginTop": "16px",
          }
        }
      >
        <withRipple(Function)
          onClick={[Function]}
          type="button"
        >
          <FormattedMessage
            id="general.cancel-btn.text"
          />
        </withRipple(Function)>
        <withRipple(Function)
          raised={true}
          type="submit"
        >
          <FormattedMessage
            id="general.ok-btn.text"
          />
        </withRipple(Function)>
      </Component>
    </form>
  </Function>
</Function>
`;
